server/protocols/webadmin/webadmin-data/src/main/java/org/apache/james/webadmin/routes/SieveScriptRoutes.java

```diff
@@ -0,0 +1,159 @@
+    private final JsonTransformer jsonTransformer;
+
+    @Inject
+    public SieveScriptRoutes(SieveRepository sieveRepository, UsersRepository usersRepository, JsonTransformer jsonTransformer) {
+        this.sieveRepository = sieveRepository;
+        this.usersRepository = usersRepository;
+        this.jsonTransformer = jsonTransformer;
+    }
+
+    @Override
+    public String getBasePath() {
+        return ROOT_PATH;
+    }
+
+    @Override
+    public void define(Service service) {
+        defineAddActiveSieveScript(service);
+    }
+
+    @PUT
+    @ApiOperation(value = "Upload a new Sieve Script")
+    @Path(value = ROOT_PATH + "/{" + USER_NAME + "}/" + SCRIPT + "/{" + SCRIPT_NAME + "}")
+    @ApiResponses(value = {
+        @ApiResponse(code = HttpStatus.NO_CONTENT_204, message = "OK"),
+        @ApiResponse(code = HttpStatus.NOT_FOUND_404,
+            message = "Invalid put Sieve script for non existent user")
+    })
+    @ApiImplicitParams({
+        @ApiImplicitParam(
+            required = false,
+            paramType = "query parameter",
+            dataType = "Boolean",
+            defaultValue = "None",
+            example = "?activate=true",
+            value = "If present, automatically activating the script.")
+    })
+    public void defineAddActiveSieveScript(Service service) {
+        service.put(USER_SCRIPT_PATH, this::addActiveSieveScript, jsonTransformer);
+    }
+
+    private Object addActiveSieveScript(Request request, Response response) throws UsersRepositoryException, QuotaExceededException, StorageException, ScriptNotFoundException {
+        User user = extractUser(request);
+        ScriptName script = extractScriptName(request);
+        sieveRepository.putScript(user, script, extractSieveScriptFromRequest(request));
+        if (isActivate(request.queryParams(ACTIVATE_PARAMS))) {
+            sieveRepository.setActive(user, script);
+        }
+        response.status(HttpStatus.NO_CONTENT_204);
+        return Constants.EMPTY_BODY;
+    }
+
+    private User extractUser(Request request) throws UsersRepositoryException {
+        return Optional.ofNullable(request.params(":" + USER_NAME))
+            .filter(Throwing.predicate(userName -> !Strings.isNullOrEmpty(userName) && usersRepository.contains(userName)))
+            .map(User::fromUsername)
+            .orElseThrow(() -> throw404withInvalidArgument("Invalid put Sieve script for non existent user"));
+    }
+
+    private ScriptName extractScriptName(Request request) {
+        return Optional.ofNullable(request.params(":" + SCRIPT_NAME))
+            .map(String::trim)
+            .filter(scriptName -> !Strings.isNullOrEmpty(scriptName))
+            .map(ScriptName::new)
+            .orElseThrow(() -> throw404withInvalidArgument("Invalid Sieve script name"));
+    }
+
+    private ScriptContent extractSieveScriptFromRequest(Request request) {
+        String script = request.body();
+        Preconditions.checkNotNull(script);
```

Comment:
how about emtpy check? BTW I think `request.body()` always return a Non-Null String. When the body is empty or not what expected, we could halt an error with a explained message to client.